### ISSUE TITLE
Evaluate IPM integration in extreme climate

### ISSUE THREAD

**Issue by kunstler** [kunstler/FUNDIV_sAPROPOS#43]
To check that my implementation of Dylan trick to integrate the IPM work well even when we predict growth and survival in extreme climate (min and max of SGDD sum of growing degree days of the species) I compared the sum of one size at t of the P matrix with the survival (should match if we integrate well growth).

To summarise what the function `mk_P_intMulti_Diag` does. It simply does a two-dimensional integration with a Gauss-Legendre integration as described in the IPM book (level 3 in one direction and 17 on the other). This done only close to the diagonal as the rest of the matrix is ~ 0 for slow-growing trees. For the rest, the mid bin method is used. (so between the black lines below)

![image](https://user-images.githubusercontent.com/4505188/37786546-445187be-2dfd-11e8-9969-defe69643b72.png)

Here is in an example for *Fagus sylvatica* that seems to work well (red lines are the probability of survival and black points sum of P matrix for that size). 

![image](https://user-images.githubusercontent.com/4505188/37786540-404cb9ea-2dfd-11e8-9ac2-7bf390f870b4.png)

There is just some eviction left for size close to the max (just because the growth kernel predicts growth outside the range). I'm not sure if this is a problem or not and if we should correct for that (by adding the missing probability to the last size as it is done in IPMpack). @dzchilds and @robcito What do you think of that? Is it ok should I try to solve that further?

**Comment by kunstler** [comment]
For the eviction pb at very large sizes, I implemented the two options of IPMpack (constant and sizeExtremes). The constant methods give the blue points (vs black points for un-corrected for eviction):

![selection_003](https://user-images.githubusercontent.com/4505188/37913289-d03df888-3114-11e8-856b-e73d575e9bc5.png)

I'm not sure if the eviction is really problematic in our case. It is always occurring way above the mean size at death and generally above the 95% CI. Williams et al. (2012) provide methods to evaluate whether the eviction is an issue for lambda but with a full IPM (including fecundity). Not sure what is the best way to adapt that in our case where we don't have fecundity recruitment. 



Williams, Jennifer L., Tom EX Miller, and Stephen P. Ellner 2012. Avoiding Unintentional Eviction from Integral Projection Models. Ecology 93(9): 2008–2014.

**Comment by dzchilds** [comment]
My guess is that it might be a bit dangerous to use mean size at death with trees as a reference size because you can end up with a long tail at the "large, high survival" end of the distribution. These might not actually impact the means size at death much, but could be important for bits of the demography. As a quick and dirty check, I'd project a cohort forward with the growth-death kernel under the most favourable conditions (if using an environmentally-dependent kernel). If the upper size tail decays well away to ~0 density before the size at which eviction starts to kick in, you're probably safe. To be sure, it should be possible to apply the $d\lambda$ trick that Steve used in the Williams paper to any demographic metric you derive from the $P$ matrix. It's just a matter of making the two iteration matrices -- the original model and an expanded version that adds a discrete state for large individuals -- then calculating the difference in the matrix popping out of the two models.

**Comment by kunstler** [comment]
Ok thank you! I will plot the survivorship curves for all species to see if eviction occurs in size range problematic or not. If there are some problems I will look into adapting the code of Williams paper.

**Comment by dzchilds** [comment]
Yep, sounds like a sensible plan. It should be OK to extend the upper size range -- perhaps using the empirical size distribution to determine a consistent, per species value -- and cap growth with a ceiling at this size.

**Comment by kunstler** [comment]
In the function `mk_P_e_intMulti_Diag` I have implemented 3 different correction of eviction:

1. "sizeExtremes" (based on IPMpack) where all evicted probability are dumped in the maximum size class.

2 "constant" where the evicted probability is divided equally across size class.

3. "ceiling" as in Williams et al. by creating an additional size class for tree larger than ceiling size. 

Then I compared the derivation of lifespan and size at death with the different methods. The effect is generally small except for some species for which size at death can be very different when correction for eviction with the ceiling method. Not sure why (and if this is done properly).
